src/Monitor/Tracing/Zipkin.hs

```diff
@@ -140,26 +129,55 @@ new (Settings mbHostname mbPort mbEpt mbMgr mbPrd ignore) = liftIO $ do
       , HTTP.port = maybe 9411 fromIntegral mbPort
       , HTTP.requestHeaders = [("content-type", "application/json")]
       }
-  void $ let prd = fromMaybe 0 mbPrd in if prd <= 0
-    then pure Nothing
-    else fmap Just $ forkIO $ forever $ do
-      threadDelay (microSeconds prd)
-      let flush = flushSpans mbEpt tracer req mgr -- Manager is thread-safe.
-      if ignore
-        then catch flush (\(_ :: HttpException) -> pure ())
-        else flush
-  pure $ Zipkin mgr req tracer mbEpt
+    zpk = Zipkin mgr req tracer mbEpt
+  for_ (microSeconds <$> mfilter (> 0) mbPrd) $ \delay -> fork $ forever $ do
+    threadDelay delay
+    let handler exc = 
+          case fromException exc of
+            (Just (PublishFailed missedSpans _)) -> refillQueue (zipkinTracer zpk) missedSpans
+            Nothing                    -> unless ignore $ throw exc
+    publish zpk `catch` handler
+  pure zpk
+
+refillQueue :: (MonadIO m, MonadBaseControl IO m) => Tracer -> [ZipkinSpan] -> m ()
+refillQueue tracer = 
+  let unspan (ZipkinSpan _ sample) = sample
+      queue = spanSamples tracer
+      fillOne item = void . atomically $ do
+        cannotPush <- isFullTBQueue queue
```

Comment:
This check is very important, lest an uncaught exception is raised. Of course, as a result, if the queue is full, there will be some traces lost (which is why I set a fairly high limit to the queue).